bot/exts/utils/attachment_pastebin_uploader.py

```diff
@@ -0,0 +1,144 @@
+from __future__ import annotations
+
+import re
+
+import aiohttp
+import discord
+from discord.ext import commands
+from pydis_core.utils import paste_service
+
+from bot.bot import Bot
+from bot.constants import Emojis
+from bot.log import get_logger
+
+log = get_logger(__name__)
+
+PASTEBIN_UPLOAD_EMOJI = Emojis.check_mark
+DELETE_PASTE_EMOJI = Emojis.trashcan
+
+
+class EmbedFileHandler(commands.Cog):
+    """
+    Handles automatic uploading of attachments to the paste bin.
+
+    Whenever a user uploads one or more attachments that is text-based (py, txt, csv, etc.), this cog offers to upload
+    all the attachments to the paste bin automatically. The steps are as follows:
+    - The bot replies to the message containing the attachments, asking the user to react with a checkmark to consent
+        to having the content uploaded.
+    - If consent is given, the bot uploads the contents and edits its own message to contain the link.
+    - The bot DMs the user the delete link for the paste.
+    - The bot waits for the user to react with a trashcan emoji, in which case the bot deletes the paste and its own
+        message.
+    """
+
+    def __init__(self, bot: Bot):
+        self.bot = bot
+        self.pending_messages = set[int]()
+
+    @staticmethod
+    async def _convert_attachment(attachment: discord.Attachment) -> paste_service.PasteFile:
+        """Converts an attachment to a PasteFile, according to the attachment's file encoding."""
+        encoding = re.search(r"charset=(\S+)", attachment.content_type).group(1)
+        file_content = (await attachment.read()).decode(encoding)
+        return paste_service.PasteFile(content=file_content, name=attachment.filename)
+
+    @commands.Cog.listener()
+    async def on_message_delete(self, message: discord.Message) -> None:
+        """Allows us to know which messages with attachments have been deleted."""
+        self.pending_messages.discard(message.id)
+
+    @commands.Cog.listener()
+    async def on_message(self, message: discord.Message) -> None:
```

Comment:
This whole method is too long imo, and would benefit from being split up.

Reply:
I've written a new method that handles both instances of waiting for the user to do a certain reaction. I don't see any other opportunity for modularization that I think would actually improve comprehendability.

Reply:
I agree with Joe, as I commented previously on to another version of this PR.

I think you can refactor out the two wait for reaction blocks into helper methods, and that lets you move those two inline functions to the very top of their respective method and murder a buncha birds with the same stone.

basically take the code that starts with the inline function declaration and ends with where you use that function, and that's a pretty reasonable scope for a helper method.